python/ray/tune/execution/ray_trial_executor.py

```diff
@@ -732,37 +820,106 @@ def has_resources_for_trial(self, trial: Trial) -> bool:
-            self._pg_manager.reconcile_placement_groups(trials)
-            self.last_pg_recon = time.time()
 
-        self._pg_manager.cleanup()
+    def _count_staged_resources(self):
+        counter = Counter()
+        for trial in self._staged_trials:
+            resource_request = trial.placement_group_factory
+            counter[resource_request] += 1
+        return counter
+
+    def _cleanup_cached_actors(self, force_all: bool = False):
+        """Clean up unneeded cached actors.
+
+        Ray Tune caches actors for re-use to avoid initialization overhead. This is
+        useful in two situations: a) to avoid scheduling overhead of actors when
+        trials run for a short time (e.g. < 1 second), and b) to avoid setup overhead
+        when trials initialize a heavy environment/dependencies (e.g. rllib).
+
+        Actors are cached when a trial is stopped. However, at that point in time
+        we don't always know if we will need the actor later or not. E.g. if all
+        subsequent trials have different resource requirements, we wouldn't need to
+        cache the actor.
+
+        Tune will generate a new trial in the next iteration of step(). Only once
+        this new trial is generated can we know if we will actually use the cached
+        actor soon. This is especially the case when we only run one trial at the time
+        and don't allow more pending trials: At the point of caching there would be
+        _no_ pending trial, so we would never cache the actor.
+
+        To circumvent this problem, we always cache an actor once the trial is
+        gracefully stopped (and when `reuse_actors=True`). We add this trial to
+        `self._newly_cached_actors` to prevent cleanup in the same iteration.
+
+        In the next iteration, we may have created a new trial. Then, when the step
+        ends, we know if the actor we cached in the previous iteration is needed soon.
+        This is checked in this very method. If it is not needed, it will be cleaned
+        up (as it will have been removed from `self._newly_cached_actors` by now).
+
+        This method fetches the required resources for all pending trials and the
+        resources for all cached actors. If we cached more actors than we need, we
+        terminate the excess actors and free the resources.
+        """
+        if not force_all and time.time() - self._last_cached_actor_cleanup < 0.5:
+            return
+
+        staged_resources = self._count_staged_resources()
+
+        for resource_request, actors in self._cached_resources_to_actor.items():
+            re_add = []
+
+            while len(actors) > staged_resources.get(resource_request, 0) or (
+                force_all and len(actors)
+            ):
+                actor, allocated_resources = actors.pop()
+                if actor in self._newly_cached_actors and not force_all:
+                    self._newly_cached_actors.remove(actor)
```

Comment:
hmmm, I am trying to think what is the invariant here. 
The current logic (which is based on if the actor is newly cached, and this only spares them from being cleaned up for one tune loop or whatever the timer says, 0.5s for now), doesn't seem to work well heterogeneous resource types, does it?
It seems, mostly likely, there will only be one more trial added to be pending in the next tune loop. This means only one cached actor of such resource type is meaningfully reused. All the others will be just discarded. Is that efficient?
Should we just do everything time based? We can clearly say that there is a grace period of xxx seconds for each resource type to allow for exact one newly cached actor to be spared from being cleaned up. 
i.e. 
len(cached_actor) <= num_staged + 1, for each resource type.

Reply:
See above and offline discussion :-)